### PR TITLE
feat: added BOX doctype, Shipping Information as child table in Item

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
Ref: [Asana](https://app.asana.com/0/1192407896547052/1199520543546492)

### CODE REVIEW

File: erpnext/stock/doctype/item/item.js [hunk 1]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
 	}
 })
 
+frappe.ui.form.on("Shipping Information", {
+	box_capacity_for_item: function(frm, cdt, cdn) {
+		var row = locals[cdt][cdn];
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
		let row = locals[cdt][cdn];
```

File: erpnext/stock/doctype/item/item.js [hunk 2]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
...
+		if (row.box_capacity_for_item > row.max_box_capacity) {
+			frappe.throw(__("Box Capacity for Item cannot be greater than Maximum Box Capacity"))
+		}
+	}
+})
+
+cur_frm.fields_dict['shipping_information'].grid.get_field('uom').get_query = function(doc,cdt,cdn) {
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
frm.set_query('shipping_information', 'uom', function(doc,cdt,cdn) {
```

Move this to refesh trigger

File: erpnext/stock/doctype/item/item.js [hunk 3]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
...
+	var row = locals[cdt][cdn];
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
	let row = locals[cdt][cdn];
```

File: erpnext/stock/doctype/item/item.js [hunk 4]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
...
+	let uom = doc.shipping_information.map(info=> info.uom);
+	let box = doc.shipping_information.map(info=> info.box);
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
	let uom = doc.shipping_information.map(info => info.uom);
	let box = doc.shipping_information.map(info => info.box);
```

File: erpnext/stock/doctype/item/item.js [hunk 5]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
...
+	box = box.splice(0, box.length-1)
+	if(box.includes(row.box)) {
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
	if (box.includes(row.box)) {
```

File: erpnext/stock/doctype/item/item.js [hunk 6]
```diff
@@ -812,6 +812,29 @@ frappe.ui.form.on("Item Supplier", {
...
+		return {
+			filters:{
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.js]:
```suggestion
			filters : {
```

File: erpnext/stock/doctype/packing_slip/packing_slip.js [hunk 7]
```diff
@@ -9,12 +9,41 @@ frappe.ui.form.on('Packing Slip', {
 	}
 });
 
+frappe.ui.form.on('Packing Slip Item', {
+	box_type: function(frm, cdt, cdn) {
+		var row = locals[cdt][cdn];
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.js]:
```suggestion
		let row = locals[cdt][cdn];
```

File: erpnext/stock/doctype/packing_slip/packing_slip.js [hunk 8]
```diff
@@ -9,12 +9,41 @@ frappe.ui.form.on('Packing Slip', {
...
+		if (row.box_type && row.qty) {
+			frappe.call({
+				method: "erpnext.stock.doctype.packing_slip.packing_slip.fetch_no_of_boxes_required",
+				args: {
+					"item_code": row.item_code,
+					"box_type": row.box_type,
+					"qty": row.qty
+				},
+				callback: function(r) {
+					if (r.message) {
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.js]:
```suggestion
					if (r && r.message) {
```

File: erpnext/stock/doctype/packing_slip/packing_slip.js [hunk 9]
```diff
@@ -9,12 +9,41 @@ frappe.ui.form.on('Packing Slip', {
...
+						frappe.model.set_value(cdt, cdn, "no_of_boxes_required", r.message)
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.js]:
```suggestion
						frappe.model.set_value(cdt, cdn, "no_of_boxes_required", r.message);
```

File: erpnext/stock/doctype/packing_slip/packing_slip.js [hunk 10]
```diff
@@ -9,12 +9,41 @@ frappe.ui.form.on('Packing Slip', {
...
+					}
+				}
+			})
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.js]:
```suggestion
			});
```

File: erpnext/stock/doctype/packing_slip/packing_slip.js [hunk 11]
```diff
@@ -9,12 +9,41 @@ frappe.ui.form.on('Packing Slip', {
...
 	}
 }
 
+cur_frm.fields_dict['items'].grid.get_field('box_type').get_query = function(doc, cdt, cdn) {
+	var row = locals[cdt][cdn];
+	return {
+		query: "erpnext.stock.doctype.packing_slip.packing_slip.get_box_type_for_item",
+		filters:{ 'item_code': row.item_code}
+	}
+}
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.js]: Change this to one suggested above

File: erpnext/stock/doctype/packing_slip/packing_slip.py [hunk 12]
```diff
@@ -184,3 +185,33 @@ def item_details(doctype, txt, searchfield, start, page_len, filters):
...
+		Fetches boxes available in child table Shipping Information from Item master.
+	"""
+	return frappe.db.sql("""
+		SELECT
+			box
+		FROM
+			`tabShipping Information`
+		WHERE
+			parent = (%s)
+	""",filters.get("item_code"))
```
Comment (hrwX) [inline erpnext/stock/doctype/packing_slip/packing_slip.py]: Use ORM

**Review by hrwX — CHANGES_REQUESTED** [review]

**Comment by rvpasari** [comment]
Waiting on a sign off on the functionality from Kenneth before I approve the PR.

**Review by rvpasari — DISMISSED** [review]

**Review by rvpasari — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**